### ISSUE TITLE
Custom community signals for accounts & assets in Ocean Market. For KYC, rugpullindex, ..

### ISSUE THREAD

**Issue by trentmc** [oceanprotocol/market#252]
Summary: give the community tools to self-curate (a) accounts and (b) assets with community-created signals. The signals could be positive-leaning, negative-leaning, or purely informational.

Inspiration: "add custom token" in Metamask, Balancer, Uniswap. Browser plug-ins.

Examples to incorporate:
* Account, Binary signal: "Rug puller y/n according to realdatawhale"
* Account, Binary signal with link: "Was KYC'd by Fractals, <link w more info>". ([Related issue](https://github.com/oceanprotocol/multi-repo-issue/issues/39))
* Asset, Continuous-valued signal: "Tim's rug pull index value" 
* Asset, Binary signal: "Has certificate issued by authority x"

The info for each would come from an API run by a community member. Above, it's realdatawhale, Tim, and Fractals. Evolution will drive towards great signals over time. It could get increasingly sophisticated. But no extra effort from us, once we have the foundation.

What info could it be? Possible pieces of info:
* structured JSON blob, or arbitrary JSON blob
* a super-short string, or a long string
* a url
* a image
* a color
* location (see below)

Q: where to put it? Possible places:
* Like our past badges
* In a custom "statistics" or "community info" section 

Q: how to store ways for user to save the set of custom badges that they've enabled, without OPF doing its own storage or custody of that. Options include: use 3box 'data vault' functionality, extend DDO.sol for user info (and have encryption), a custom on-chain solution, more.

**TODOs for this issue:**
* [ ] Refine the spec
* [ ] Tech spike(s) to explore remaining Q's
* [ ] UI reflects the spec

**Comment by trentmc** [comment]
Here's an example from "Farm Ocean" Telegram channel: people are already making their own badges. This will help to reinforce the existing community efforts.

 ![]<img src="https://images.zenhubusercontent.com/5df770bc5c7a2019891bc4a0/8c278d08-8213-4843-87a2-057ae9a676c8" width="40%"> ![]<img src="https://images.zenhubusercontent.com/5df770bc5c7a2019891bc4a0/6e29fcbe-1eae-4ac7-b335-6b36d1edcbc5" width="40%">

**Comment by anapheys** [comment]
Overall goal: mitigate rug pulls, mitigate IP violation, and more.
Strategy here: give the data set/pool publisher the ability to deposit their LP tokens into a "data publisher DAO" to earn a badge.

Data set/pool publishers could receive a badge if all of their LP tokens were deposited into a smart contract from which they could exit anytime. Depositing LP tokens into the "data publishers DAO" signals commitment to the pool and is transparent if they decide to pull them out. [One cannot rug pull if the tokens are in a different smart contract]. If you pull LP tokens out of "data publishers DAO" then you lose your data commitment badge. The rules of the Data Committers are defined by the "agreement" of the "data publishers dao".

**Comment by kremalicious** [comment]
I do not see this as high priority at all. It will only lead to a cluttered interface with no real value and the mechanisms for adding those badges. And the mechanisms in the background seem to require new components, which for the feature I don't think it's worth the hassle

**Comment by trentmc** [comment]
> I do not see this as high priority at all. It will only lead to a cluttered interface with no real value and the mechanisms for adding those badges
Disagree. So far it seems to be the best "open" means to add subjective signals, and signals that the community deems important. All which help towards discovery, and answering Ocean Market user Q's like "should I stake on this dataset? Buy it?" etc. Which means that you and I need to discuss further.

**Comment by trentmc** [comment]
Note: [Multi-repo #32](https://github.com/oceanprotocol/multi-repo-issue/issues/32) "People who have consumed data can give a rating" is a bit related to this issue. I see that it's much easier to implement (and less controversial) if treated as a community signal. See especially [this comment](https://github.com/oceanprotocol/multi-repo-issue/issues/32#issuecomment-757707327)

**Comment by SarahKay99** [comment]
Hi everyone. I'm coming here from Multi-repo #32. I think community-created signals are a good idea but there are just a few problems:

- There could be too many community ratings systems. The real important ones will lose their value as the users become overwhelmed with ratings to check.
- In the instance there are too many ratings signals: How do we judge which ratings should be shown first? 
- In the instance there are two few ratings signals: Ratings will become centralized if we rely on APIs.
- Some ratings are objectively more valuable than others, i.e. the Rug Pull Index is more valuable than someone's personal opinion. This is kinda the same as problem #2: how do we judge?

Still I think it is an awesome idea to let community members participate in monitoring the market. I think this community badge/community ratings API system could work in a properly decentralized way in a few years. For now the centralized APIs are a decent solution, so long as their centralization is temporary. 

Side note: A ratings signal which lets people share the models they trained would be very valuable. 

Not sure what "refining the spec" means, if someone can explain this a bit better to me I'd like to draw some UI/UX diagrams.

**Comment by trentmc** [comment]
Great Q's. Answers are below. Summary is that we're ok, we let the community adapt.

> There could be too many community ratings systems ... 

Actually, we want tons of community rating systems. The community will self-adapt towards using signals they like the most (e.g. via conversation on Data Farming TG channel).

> In the instance there are too many ratings signals: How do we judge which ratings should be shown first?

*We* don't. Each individual user chooses which they want, and uses it. Towards defaults for what Ocean Market shows, we will also observe what the community uses the most, and what we think makes the most sense. (And if others don't like it, they can fork and run their own version of Ocean Market.)

> In the instance there are two [sic] few ratings signals: Ratings will become centralized if we rely on APIs.

All raw data ultimately comes from a centralized location. Signals are no different. And that's ok, all data (incl. signals) are at the "leaf" nodes of the system. The center of the system -- the smart contracts and the substrate -- are decentralized. 

Second, ultimately there could be competition for providing signals. For a given signal, all else being equal, decentralized will prevail over centralized (less trust needed).

> Some ratings are objectively more valuable than others ... how do we judge

Like mentioned above, we don't, the community does. 

> Side note: A ratings signal which lets people share the models they trained would be very valuable.

This is not a ratings signal. This is simply data. Ocean Market doesn't care what type of data is being bought and sold. You can use it to sell trained models right now.

**Comment by trentmc** [comment]
> Not sure what "refining the spec" means, if someone can explain this a bit better to me I'd like to draw some UI/UX diagrams.

When I think about "refining the spec" I think about it as going through the levels in "Elements of User Experience". [Single image](https://uxcollection.com/wp-content/uploads/2018/03/elements.gif). [PDF overview](http://www.jjg.net/elements/pdf/elements_ch02.pdf). 

For a given issue of any level of complexity, it's useful to consider how much to handle each step, going from step (1) through to step (5).
 * GUI is step (5) - "surface plane", something to leave to the end. 
 * This ticket has already basically defined steps (1) and (2) - "strategy plane" and "scope plane". 
 * Step (3) is the next step - "structure plane" including "interaction design" and "information architecture". 
 * Then step (4) then step (5). 

I'm happy to provide feedback as you proceed. And I expect (and hope!) that @kremalicious will too:)

**Comment by SarahKay99** [comment]
Do you have existing UI/UX diagrams for the Ocean Market which we can recycle components of? @trentmc

**Comment by lopezi** [comment]
subscribing ping

**Comment by lopezi** [comment]
Here's a first draft depicting UI elements to visualise trust signals on data silos and accounts:

![trust](https://user-images.githubusercontent.com/388476/104617632-937bca00-5683-11eb-9de6-88790967bb42.png)

**Comment by trentmc** [comment]
This draft is a good start!

The summarizing "Trust Index" value has issues: 
* "Trust Index" value is a function of several signals. How to choose each? Does it change based on the user choosing different signals? How do you combine the signals into a final value? What if one or more of those signals is gamed? 
* The user may see it as green then think "I can trust this dataset now". But something bad could still happen, eg a rug pull. This is not theory, it happened with the "Data Launch Partners" badge, which is why we removed it. The point is, there should be no direct signal of "stamp of approval" because it can lead to false conclusions. The best we can do is provide signals to the user and they can choose what to do with them.
 * **There's an easy fix** on this: remove the summarizing "Trust Index" value.

Other feedback:
* the draft is only "step (5) - surface plane" of Elements of User Experience. My previous comment suggested what other things you should be covering for UX. E.g. interaction design to cover the workflows like adding a new signal. Typically the work for that *precedes* step 5.
* the draft does not show several of the things laid out in this issue Description. E.g. it should show each of:
  * Account, Binary signal: "Rug puller y/n according to realdatawhale"
  * Account, Binary signal with link: "Was KYC'd by Fractals, ". (Related issue)
  * Asset, Continuous-valued signal: "Tim's rug pull index value"
  * Asset, Binary signal: "Has certificate issued by authority x"
* the draft should show how a community member adds or removes a given signal. E.g. is it in the 'settings' screen? Or by looking at the existing signals and clicking a trash icon to delete, or a + icon at the bottom to add a new signal

Here's a checklist to help understand if you've got a good first draft done: Does it adhere to the Description of this issue? Does it document the steps of Elements of User Experience (not just step 5)? Does it address the concerns I've raised so far?

Thanks again, this is a good start, let's keep iterating:)

**Comment by lopezi** [comment]
Which "Trust Index"? Gone.

I don't understand the specs from the sub.items when you mention:
- _the draft does not show several of the things laid out in this issue Description. E.g. it should show each of:

My interpretation is that, e.g. _Account, Binary signal: "Rug puller y/n according to realdatawhale"_ 
was a binary input with name _Rug puller_ represented by a _yes/no:_ output on the interface; accordingly to the UI draft above.

Pragmatically, for me to build the UI for this feature I would like to have:

- The exact name of the parameters to visualize
- - are this parameters coming from the backend/api, or are "random" user inputs? (1)
- The corresponding value each parameter has
- who adds the signals? everyone? who removes them? anyone? 

(1) you mention "the draft should show how a community member adds or removes a given signal."

Real use cases and real data will help continuing the process. Thanks!

**Comment by trentmc** [comment]
> Which "Trust Index"? Gone

Great:)

> My interpretation is that, e.g. Account, Binary signal: "Rug puller y/n according to realdatawhale"
> was a binary input with name Rug puller represented by a yes/no: output on the interface; accordingly to the UI draft above.

That's the correct interpretation. The draft above does indeed have that; good. 

There are other signal types to handle; "Account, Binary Signal" is one type. I listed other types, in my previous comment as well as in the description. E.g. Tim's rug pull index isn't just a "yes/no" but has a continuous value. Therefore it's "Asset, Continuous-Valued Signal".

> Pragmatically, for me to build the UI for this feature I would like to have:
> The exact name of the parameters to visualize - are this parameters coming from the backend/api, or are "random" user inputs? 

They are values coming from user-supplied services as REST APIs. They can follow the same pattern as is used for purgatory, or for "launch partner" (the latter is deprecated). The exact signal can be quite narrowly defined, to handle "Account, Binary Signal" and the other types. 

The signal is shown e.g. like your prototype has, but more than just "Account, Binary signal" type.

Anyone in the community can set up their own REST API service, and share the url with the community. (It could simply be in TG to start. Or as comments to this issue. At some point there could be a registry of these. But one step at a time.)

> who adds the signals? everyone? who removes them? anyone?

The user adds and removes the signals for themselves. In a vein similar to Metamask "add custom token". The exact means is part of what needs to be determined during the design process, ie now.

> The corresponding value each parameter has
> Real use cases and real data will help continuing the process. Thanks!

These are given in the description, e.g. "Rug puller y/n according to realdatawhale". Those are real use cases, and have real data behind them.

**Comment by lopezi** [comment]
Follows the second iteration of the ui elements and ux for the custom community signals feature

![trustflow](https://user-images.githubusercontent.com/388476/105758858-a75ae200-5f47-11eb-8c52-9c1009395417.png)

**Comment by trentmc** [comment]
I like this a lot! It's come a long way from before, great stuff.

Things I like in particular:
 - signals for both assets and for accounts
 - handles both continuous-valued and binary-valued assets
 - some support for links
 - a first-cut UI for how to add the signals

Here are a few suggestions to tweak this further:
- next to "USE | POOL | TRADE", it says "TRUST". Please don't use "TRUST" label, it is a red herring and can be misleading. Perhaps "SIGNALS"? Maybe there's an even better label yet. 
-The "add custom signals" part could probably be a little less prominent. Perhaps put it at the very bottom of the panel rather than the top? Or perhaps have a small "+" symbol at the bottom of the account signals list, and another at the bottom of the asset signals list.
- each signal type might have a link, rather than a special type that is only a link with no other information. 
- it gives "stars" icon to represent a continuous-valued signal. This would only work for some signals where higher=good. But other signals might be lower=good or middle=good or something else. Perhaps just show the real-valued number?
- it gives a "medal" icon to represent a certificate which is a binary signal. That assumes true=good. But other signals might be false=good. Perhaps just show "yes" or "no" (like it does for another signal).

**Comment by lopezi** [comment]
Good feedback thanks. Follows another iteration. 

![trustflow_v2](https://user-images.githubusercontent.com/388476/106002495-1d2a8f00-60a9-11eb-9bc5-0346c541b7e5.png)

**Comment by trentmc** [comment]
Great update! You have addressed all my concerns here. 

With one exception: "each signal type might have a link, rather than a special type that is only a link with no other information" still needs addressing. Maybe I can clarify further. In the prototype above, for signal "Certificate x" which gives a value of "yes" for this asset, also have also a url for that signal that the user can click on to open another tab to get more information e.g. about how that signal is calculated. And for "Rug puller - realdata" there is also a url too. And so on, for every signal. This also means that "Certificate Y" is not just a url, it also needs to have some binary value or continuous value.

Moving forward from this, perhaps a next step is to specify what the REST API interface looks like. The starting point is the patterns used for (a) asset purgatory, (b) account purgatory, and (c) data launch partners [now deprecated].

**Comment by lopezi** [comment]
Adding links to the signals according to your feedback. 

This is not pixel perfect as I don’t have the font and sometimes I just go to the browser devtools, design things on the ocean market to grab the styles and get a print screen. A reusable design system along with the artwork files / interactive styleguide would be the ideal setup. Do you guys have one?

![trustflow_v3](https://user-images.githubusercontent.com/388476/106161950-782eb580-617f-11eb-9da9-b40da986c38f.png)

> Moving forward from this, perhaps a next step is to specify what the REST API interface looks like. The starting point is the patterns used for (a) asset purgatory, (b) account purgatory, and (c) data launch partners [now deprecated].

I am not aware of these patterns you mention or the specs for the API interface. 

Can you please provide more input?
- Where is located this REST API interface?
- What are the inputs / outputs / functionality ?
- Is it anything similar to the above rug puller example for adding signals?

**Comment by trentmc** [comment]
> This is not pixel perfect ...

That's ok. We have several templates, one will likely fit. Ocean Market uses gatsby which is quite easy to extend, if needed.

> I am not aware of these patterns you mention or the specs for the API interface.
> Can you please provide more input?

The Ocean repo [market-purgatory](https://github.com/oceanprotocol/market-purgatory) is for a RESTful API service giving a binary-valued signal for actors (accounts), and a binary-valued signal for assets. In its case, it draws on data that is stored in the Ocean repo [list-purgatory](https://github.com/oceanprotocol/list-purgatory).

There's another pattern that had been used for data launch partners. It's a binary-valued signal for actors. The info for that is this (archived) Ocean repo [list-datapartners](https://github.com/oceanprotocol-archive/list-datapartners).

An example of continuous-valued signal per asset is Tim's rug pull index. More info is at https://rugpullindex.com/. 

I would suggest to take these as example patterns of services, but then come up with a slightly more generalized REST API spec.

Finally @kremalicious do you have more guidance on the REST API interface / how this all might connect?

**Comment by lopezi** [comment]
ping

**Comment by SarahKay99** [comment]
How is the task going? 

Let me know if you need some help with the UI, or if we are ready to program. 

If you guys don't have enough people to do the programming I am happy to help. There are also some people in the Discord who I'm sure could help.

**Comment by kremalicious** [comment]
Good development in here but unfortunately none of the UI can be built until we figure out the standardization of the data sources behind it. As of right now there are simply multiple external API endpoints like rugpullindex.com with different responses and nothing which connects them.

If the intention is to build something like how UniSwap uses [token-lists](https://github.com/Uniswap/token-lists) throughout their app but for standardized "signal-lists" then well, this standard has to be created first. After that, the intended projects to get data from need to then create lists and APIs following this standard. Cause we can't deal with random API endpoints where every response's data structure is different. So this is a huge undertaking which would need to be done first, that whole UI to make use of those "signal-lists" is the easy part.

Apart from that, good UI drafts for making use of a mechanism like token-lists to add those lists in a UI element. Just its placement within the app feels a bit clunky. I do not see users going to each and every dataset and adding some "signals" only they can see. Cause there is no database so all we can do is save those user added signals in the user's browser. With a standard described above this is fine, but then it makes much more sense to have this as some global setting across the app, so whatever mechanism is used will be applied to all data sets.

But realistically, to me the only solution which could be shipped right now is to come up with UI elements for representing a response from Tim’s rugpullindex, and display it somewhere within the _Pool_ tab

**Comment by trentmc** [comment]
>  the only solution which could be shipped right now is [Tim's] ...

Besides Tim's index, we have requests for whether KYC'd. We have curation of "rug puller y/n" by realdatawhale. We have others talking about issuing "trust certificates". To summarize:
- Account, Binary signal: "Rug puller y/n according to realdatawhale"
- Account, Binary signal with link: "Was KYC'd by Fractals, ". (Related issue)
- Asset, Continuous-valued signal: "Tim's rug pull index value"
- Asset, Binary signal: "Has certificate issued by authority x"

That's four separate use cases.

> none of the UI can be built until we figure out the standardization of the data sources behind it. 

Great. So let's use this issue to drive a simple standard.

> standard like token sets...

Of course can't be a list because that is simply a Y/N. The info here requires something more general. It could be a real-valued number. Or to be even more general yet, make it a string up to x characters. And have it also provide a url so that there is more context if the reader so desires. Finally, it should be for both assets or accounts.

Here's a first cut:
- For an account query to service X (e.g. Fractal KYC service), return a JSON blob with (a) a string with info (b) a url
- For a data asset query to service Y (e.g. Tim's index), return a JSON blob with (a) a string with info (b) a url

That's it. Simple, clean. This already covers four specific use cases.

**Comment by kremalicious** [comment]
So the summary then implies implementing 4 different API calls and show their response in a nice way in `market`. That's it, nothing with adding some unknown API endpoints. There is nothing connecting those APIs. We will not find a way to unify multiple APIs without either creating an API on top of that or making everybody else create APIs based on some newly developed schema. By design, all those APIs are different, it does not make sense that e.g. rugpullindex.com has same data structure than an API about KYC status

**Comment by trentmc** [comment]
> 4 different API calls 

Here's how I see it. The market has a two lists. 
* list of "account info" service urls
* list of "data asset info" service urls

There's a good default. Each list can be refined / defined by the user. That makes 2 different types if API calls (not 4).

>  So figure out how to represent data from rugpullindex and Fractal KYC

The account info service url outputs the following:
```
{ info: "<10 char string>",
  url: "<url string>
} 
```

The data asset info service url outputs the following:  `(same as above)`

**Comment by kremalicious** [comment]
Still not entirely clear to me what all these lists are about.

Again, the issue defines to have API responses from 3 sources (datawhale, rugpullindex, Fractal) and it's an excellent idea, it's using existing projects and connecting them with the market UI to provide trust signals. So let's just do that before coming up with some aggregate mechanism for those? To me those are the `custom community signals`. No list creation or API specification creation involved, just the market using existing APIs and their data structure. Except for datawhale I guess, cause it's a data set.

But then there's things like `Has certificate issued by authority x` and `custom badges`, seems all these very different mechanisms for it are floating around:

1. user adds API endpoint URLs in the market UI
2. user adds URL to lists of API endpoints
3. user adds URL to lists of assets
4. user adds URL to lists of accounts

3. & 4. could be made similar to `list-purgatory`. As an example: there could be a `list-golden` maintained by community with defined rules on how to get on it, and it references assets by DID. In the UI this can then be used to show e.g. a gold badge on a data set. Same for accounts. Could also be made into something where only those assets in that list show up in the UI, like with the default token list on UniSwap. For this we can define the schema of those lists, as long as they are public, they'll work

But 1. & 2. will not work that easily until we have an API schema specified and then have others build their project's APIs based on that. Right now I fear we would try to design other people's APIs and right now telling rugpullindex & Fractal to rebuild their APIs.

**Comment by trentmc** [comment]
A bit more on the motivation of this issue. We get regular suggestions about new curation signals to build or incorporate. I just had another one yesterday. Of the top of my head, here are some: 3 data quality signal providers already in Ocean Market (realdatawhale is one; there are two others), 1 Tim's index, 3 KYC (Fractals, 3box KYC partner, big enterprise in talks with Bruce), 1 data quality suggestion from yesterday (via another collaborator). That's 8. And that's without even peoples' awareness that this exists, since it doesn't exist yet. So you better believe there could be a lot of these signals.

So do we handle all these signals?
1. One option to handle this is for the Ocean core team to build the signal themselves, or replicate them. This is what we're doing right now. It's a huge effort or simply not happening, and doesn't easily open for other community members' signals or a natural evolution towards the best signals
2. Another option is to leverage the existing signals and make a place where other signals can be easily incorporated. This avoids huge efforts by core team (or things simply not happening), and makes Ocean Market more open to evolution of useful signals

This ticket is about moving us from the status quo (1) to something more scalable and open-ended (2).

----

Here's what @kremalicious and I agreed on in our conversation yesterday.

## 1. Schema 

Schema for signals, by example:
```
Asset info service:
{
  "name": "Rug Pull Index",
  "value": "0.823",
  "url": "https://rugpullindex.com/$assetid"
}

Account info service:
{
  "name": "RealDataWhale Good Actor",
  "value": "Yes",
  "url": "https://realdatawhale.com/$accountid"
}

Account info service:
{
  "name": "Fractals KYC",
  "value": "Pending",
  "url": "https://fractals.com/kyc/api/$accountid"
}
```

Each signal provider would follow this schema. They're incentivized to do so because the url takes the user to their service which is a monetization opportunity for them. Eg buy the full realdatawhale dataset for much more detailed info.

The value just a string, on purpose. It's simple, flexible, and doesn't require special interpretation by the frontend code. The url is there to provide more context so that any Market user can learn more if they wish (range information, how the value is calculated). 

There wouldn't be any authorization required. 

## 2. Lists 

The market has a two lists to describe who the "community signal" service providers are.
- list of "account info" service urls
- list of "asset info" service urls

These lists are maintained in a special repo following the same pattern as `list-purgatory` repo: a json file for each list. Permissions to write to that repo would start with the Ocean core team but could be extended in time. This is not full customizability for each user, but a very good stepping stone towards that.

\ fin

**Comment by TimDaub** [comment]
Great thread. And also, IMO the ideas here could easily be converted into a distinct technical product. Mainly speaking about the methodology of broadcasting a signal and about its data model.

I think we all have a hunch that there's no canonical way of sending digital signals over HTTPS for now. In fact, a few months ago I started [writing a specification for a simple signal replication protocol over HTTPS](https://timdaub.github.io/SSRP/). But so far, I've not gotten anywhere. But this thread helps.

I'll go through soon and add more to the spec. Help is welcome if there's interest. Repo is [here](https://github.com/TimDaub/SSRP).

**Comment by TimDaub** [comment]
I just published the RPI docs for its public API: https://rugpullindex.com/docs

**Comment by kremalicious** [comment]
> IMO the ideas here could easily be converted into a distinct technical product

that's exactly what should happen instead of cranking this into the market UI. Once this distinct product is launched and works, then can be reevaluated what parts of it could benefit this market repo. Feel free to move this issue to some other repo, here it will be closed as out of scope.
